webgui-new/src/service/git-service.ts

```diff
@@ -0,0 +1,188 @@
+import thrift from 'thrift';
+import { GitDiffOptions, GitService } from '@thrift-generated';
+import { config } from './config';
+
+let client: GitService.Client | undefined;
+export const createGitClient = (workspace: string) => {
+  if (!config) return;
+  const connection = thrift.createXHRConnection(config.webserver_host, config.webserver_port, {
+    transport: thrift.TBufferedTransport,
+    protocol: thrift.TJSONProtocol,
+    https: config.webserver_https,
+    path: `${config.webserver_path}/${workspace}/GitService`,
+  });
+  client = thrift.createXHRClient(GitService, connection);
+  return client;
+};
+
+export const isRepositoryAvailable = async () => {
+  if (!client) {
+    return false;
+  }
+  try {
+    return await client.isRepositoryAvailable();
+  } catch {
+    return false;
+  }
+};
+
+export const getRepositoryList = async () => {
+  if (!client) {
+    return [];
+  }
+  try {
+    return await client.getRepositoryList();
+  } catch {
+    return [];
+  }
+};
+
+export const getRepositoryByProjectPath = async (path: string) => {
+  if (!client) {
+    return;
+  }
+  try {
+    return await client.getRepositoryByProjectPath(path);
+  } catch {
+    return;
+  }
+};
+
+export const getReferenceTopObject = async (repoId: string, branchName: string) => {
+  if (!client) {
+    return;
+  }
+  try {
+    return await client.getReferenceTopObject(repoId, branchName);
+  } catch {
+    return;
+  }
+};
+
+export const getReferenceList = async (repoId: string) => {
+  if (!client) {
+    return [];
+  }
+  try {
+    return await client.getReferenceList(repoId);
+  } catch {
+    return [];
+  }
+};
+
+export const getBranchList = async (repoId: string) => {
+  if (!client) {
+    return [];
+  }
+  try {
+    return await client.getBranchList(repoId);
+  } catch {
+    return [];
+  }
+};
+
+export const getTagList = async (repoId: string) => {
+  if (!client) {
+    return [];
+  }
+  try {
+    return await client.getTagList(repoId);
+  } catch {
+    return [];
+  }
+};
+
+export const getTag = async (repoId: string, hexId: string) => {
+  if (!client) {
+    return;
+  }
+  try {
+    return await client.getTag(repoId, hexId);
+  } catch {
+    return;
+  }
+};
+
+export const getBlameInfo = async (repoId: string, hexId: string, path: string, localModsFileId: string) => {
+  if (!client) {
+    return [];
+  }
+  try {
+    return await client.getBlameInfo(repoId, hexId, path, localModsFileId);
+  } catch {
+    return [];
+  }
+};
+
+export const getBlobContent = async (repoId: string, hexId: string) => {
+  if (!client) {
+    return '';
+  }
+  try {
+    return await client.getBlobContent(repoId, hexId);
+  } catch {
+    return '';
+  }
+};
+
+export const getBlobOidByPath = async (repoId: string, hexId: string, path: string) => {
+  if (!client) {
+    return '';
+  }
+  try {
+    return await client.getBlobOidByPath(repoId, hexId, path);
+  } catch {
+    return '';
+  }
```

Comment:
This will be a general remark for the complete codebase.
I see that in the service layer you basically hide all errors. If anything goes wrong a default empty string, empty array or a false value is returned.

I think some kind of better error handling should be implemented: there is no much point for exceptions if we are simply catching them all and do nothing. Of course the webapp should not crash if e.g. a network request failed due to some network outage, but the user should be notified in some way that the requested data was e.g. not loaded and returned empty result is not the real result.

Reply:
Yes, this was only a temporary solution to avoid crashes, which I wanted to address later. The `react-toastify` package (https://www.npmjs.com/package/react-toastify) could be used to show an error message on the page for the user, while logging the error message to the console using `console.error`.

Reply:
That sounds good!